src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/RefAPISource.java

```diff
@@ -61,40 +73,66 @@ public static RefAPISource getRefAPISource() {
      * @param pipelineOptions options needed for credentials to produce a Genomics class for the API call.
-     * @param referenceName the "name of the reference, e.g., EOSt9JOVhp3jkwE.
-     * @return returns a mapping from reference name to String id.
+     * @param referenceSetID the ID of the reference set, e.g., EOSt9JOVhp3jkwE.
+     * @return returns a mapping from reference name to String ID.
      */
-    public static Map<String, String> buildReferenceNameToIdTable(final PipelineOptions pipelineOptions, final String referenceName) {
-        Genomics genomicsService = createGenomicsService(pipelineOptions);
-        final ReferenceSet referenceSet;
-        try {
-            referenceSet = genomicsService.referencesets().get(referenceName).execute();
-        }
-        catch ( IOException e ) {
-            throw new UserException("Could not load reference set for reference name " + referenceName, e);
-        }
-        final Map<String, String> referenceNameToIdTable = new HashMap<>();
-
-        try {
-            for ( final String referenceID : referenceSet.getReferenceIds() ) {
-                final Reference reference = genomicsService.references().get(referenceID).execute();
-                referenceNameToIdTable.put(reference.getName(), reference.getId());
-            }
-        }
+    public Map<String, String> buildReferenceNameToIdTable(final PipelineOptions pipelineOptions, final String referenceSetID) {
+        Map<String, Reference> referenceMap = buildReferenceNameToReferenceTable(pipelineOptions, referenceSetID);
+        return buildReferenceNameToIdTableFromMap(referenceMap);
+    }
 
-        catch ( IOException e ) {
-            throw new UserException("Error while looking up references for reference set " + referenceSet.getId(), e);
-        }
+    /**
+     * Given a reference set ID, get a SAMSequenceDictionary that lists its contigs and their length.
+     * Note that this does NOT use the referenceContig order from the API. Instead, we'll sort them
+     * and try to match the read sequence dictionary order, if given.
+     *
+     * @param pipelineOptions - are used to get the credentials necessary to call the Genomics API
+     * @param referenceSetID - the ID of the reference set to use
+     * @param optReadSequenceDictionaryToMatch - (optional) the sequence dictionary of the reads, we'll match its order if possible.
+     * @return a SAMSequenceDictionary that lists the referenceset's contigs and their length.
+     */
+    public SAMSequenceDictionary getReferenceSequenceDictionary(final PipelineOptions pipelineOptions, String referenceSetID, final SAMSequenceDictionary optReadSequenceDictionaryToMatch) {
+        Map<String, Reference> referenceMap = buildReferenceNameToReferenceTable(pipelineOptions, referenceSetID);
+        return getReferenceSequenceDictionaryFromMap(referenceMap, optReadSequenceDictionaryToMatch);
+    }
```

Comment:
This `getReferenceSequenceDictionary()` calls into some very sketchy-looking code, and definitely needs comprehensive unit tests.